### DISCUSSION TITLE
URIs Execute Twice

### DISCUSSION THREAD

**Discussion by wallysaurus** [Taitava/obsidian-shellcommands#404] [Debugging and testing]
When pressing the execute button directly in the settings menu, I have no problem running the script once like it should, however, when I click on a URI link to execute the command, two instances get created.

https://github.com/Taitava/obsidian-shellcommands/assets/69739715/84687073-72b1-4822-ac74-47cae127f058

**Comment by Taitava** [comment]
What is your Obsidian version?

Unfortunately I'm not able to reproduce this on my Windows 10 machine. I tried to create a link in a title of a card in a canvas file. The link executes a CMD.EXE command, and it executed correctly just once.

**Reply by wallysaurus** [reply]
I'm on the latest. (v1.5.12). Tested on both my windows desktop and mac laptop. Both do the same thing.

**Reply by wallysaurus** [reply]
Another issue I'm running into is the shell executor cannot access my installed python libraries while executing from terminal can? I'm storing my vault on iCloud currently, so that might be a problem? Although the pwd of my terminal is also in iCloud, so I'm not sure why that would be a problem.
https://github.com/Taitava/obsidian-shellcommands/assets/69739715/bcdd6879-1734-41a2-b0bc-5529ce17737a

**Reply by Taitava** [reply]
Sorry, I still can't reproduce the double execution issue. I tried now on a Linux machine (I don't have a Mac) and it only executed once when clicking a link in canvas.

- Have you tried pasting the link (just the URI) e.g. to a web browser's address bar? If you execute the shell command that way, does it execute once or twice?
- In Obisidian, can you try putting the link to a traditional markdown note file (instead of a canvas file)? Does it execute twice from there, too?

If you happen to know TypeScript and Obsidian plugin development and if you want, you can try to debug this yourself. [`SC_Plugin.registerURIHandler()`](https://github.com/Taitava/obsidian-shellcommands/blob/0e74d77d6164f6c5ff7542b520ae1db34b81fdbb/src/main.ts#L454-L532) contains the part that initially receives an URI call from Obsidian and starts an execution process. You could e.g. do `new Notice("Hello")` at line 457 to see how many times Obsidian does the URI call when you click the link.

**Reply by Taitava** [reply]
> Another issue I'm running into is the shell executor cannot access my installed python libraries while executing from terminal can? I'm storing my vault on iCloud currently, so that might be a problem? Although the pwd of my terminal is also in iCloud, so I'm not sure why that would be a problem. https://github.com/Taitava/obsidian-shellcommands/assets/69739715/bcdd6879-1734-41a2-b0bc-5529ce17737a

Can you please start a new discussion in [Q & A](https://github.com/Taitava/obsidian-shellcommands/discussions/categories/q-a) as this seems to be a separate problem? 🙂 Thank you! 👍
